### DISCUSSION TITLE
Creating a pty.

### DISCUSSION THREAD

**Discussion by uwejan** [wezterm/wezterm#3465] [General]
I am having trouble understanding, which parts am missing here.
AFAIK, is that in order to create a pty in unix generally, you will need a master and slave, comunicating with eachother functioning as a terminal session eventually.
My goal is to create a pty, and keep it open for reads/writes, till i request to destroy it later on.

```
fn main() {
    use portable_pty::{CommandBuilder, PtySize, native_pty_system, PtySystem};
    use anyhow::Error;

// Use the native pty implementation for the system
    let pty_system = native_pty_system();

// Create a new pty
    let mut pair = pty_system.openpty(PtySize {
        rows: 24,
        cols: 80,
        // Not all systems support pixel_width, pixel_height,
        // but it is good practice to set it to something
        // that matches the size of the selected font.  That
        // is more complex than can be shown here in this
        // brief example though!
        pixel_width: 0,
        pixel_height: 0,
    }).unwrap();

// Spawn a shell into the pty
    let cmd = CommandBuilder::new("bash");
    let mut child = pair.slave.spawn_command(cmd).unwrap();

// Read and parse output from the pty with reader
    let mut reader = pair.master.try_clone_reader().unwrap();

// Send data to the pty by writing to the master
    pair.master.take_writer().unwrap().write("ls -l\r\n\x04".as_bytes()).unwrap();
    //writeln!(pair.master, "ls -l\r\n\x04").unwrap();
    println!("{}", child.wait().unwrap());

    let mut out = String::new();
    reader.read_to_string(&mut out).unwrap();
    println!("{}", out);
}
```
In this example I am trying to figure out what the crate has to offer, running this will output nothing, which I also struggling to understand why.
Any pointers are appreciated. Thank you in advance.

**Comment by wez** [comment]
I'd suggest starting with https://github.com/wez/wezterm/blob/main/pty/examples/whoami.rs
When working with subprocesses, you need to take care to avoid blocking in read or write.
There are no guarantees about the ordering or fragmentation of the reads/writes.
I recommend processing the read and writes on separate threads in a real application, and also a separate thread for waiting on the child status.
